keras/src/layers/rnn/rnn.py

```diff
@@ -382,6 +382,20 @@ def call(
                 initial_state = self.get_initial_state(
                     batch_size=ops.shape(sequences)[0]
                 )
+
+        # Check for batch size mismatch when stateful
+        if self.stateful:
+            actual_batch_size = ops.shape(sequences)[0]
+            expected_batch_size = (self.states[0].shape[0] 
```

Comment:
This is the part that fails in the tests, I think `states` can be a nested structure.

You'll have to use `tree.flatten(self.states)[0]` instead of `self.states[0]`.